### ISSUE TITLE
patch_based learning : how to make validation set

### ISSUE THREAD

**Issue by 9B8DY6** [fepegar/torchio#365]
this is my code for validation_set
```
patch_size = 12
samples_per_volume = 10
max_queue_length = 300
training_batch_size = 32
validation_batch_size = 32


patches_training_set = torchio.Queue(
    subjects_dataset=training_set,
    max_length=max_queue_length,
    samples_per_volume=samples_per_volume,
    sampler=torchio.data.UniformSampler(patch_size),
    num_workers=0,
    shuffle_subjects=True,
    shuffle_patches=True,
)

patches_validation_set = torchio.Queue(
    subjects_dataset=validation_set,
    max_length=max_queue_length,
    samples_per_volume=samples_per_volume,
    sampler=torchio.inference.GridSampler(patch_size),
    num_workers=0,
    shuffle_subjects=False,
    shuffle_patches=False,
)

training_loader = torch.utils.data.DataLoader(
    patches_training_set, batch_size=training_batch_size)

validation_loader = torch.utils.data.DataLoader(
    patches_validation_set, batch_size=validation_batch_size)
```

Here is training part...

```
count = 0
loss_valid = np.array([])
iteration_list = []
loss_train = np.array([])

for epoch in range(num_epochs):
    count += 1
    i=0
    l_epoch=0
    
    for batch in training_loader:
        model.to(device).train()
        inputs_train = batch['mri'][DATA].to(device)
        target_train = (batch['age']/100.).float().to(device)
        i=i+1
        optimizer.zero_grad()
        outputs = model(inputs_train)
        target_train= target_train.reshape(-1,1)
        loss = error(outputs, target_train).to(device)
        l_epoch+=loss
        loss.backward()
        optimizer.step()
        
    loss_train = np.append(loss_train,l_epoch.cpu().detach().numpy()/i)
    print('epoch_train : ',count)
    print('train loss: ',loss_train[-1])
    print('outputs : ',outputs[:5])
    print('label : ',target_train[:5])
    torch.save(model.state_dict(), f'v0_patch_T1_reg_PD_new_1.pth')
        
    i=0
    l_epoch=0
    with torch.no_grad():
        for batch in validation_loader:
          model.to(device).eval()
          inputs_test = batch['mri'][DATA].to(device)
          target_test = (batch['age']/100.).float().to(device)
          i=i+1
          outputs = model(inputs_test)
          target_test= target_test.reshape(-1,1)
          loss = error(outputs, target_test).to(device)
          l_epoch+=loss
        loss_valid = np.append(loss_valid,l_epoch.cpu().detach().numpy()/i)
        scheduler.step(loss_valid[-1])
    
    print('valid_epoch: {}  valid_loss: {} '.format(count, loss_valid[-1]))
    plt.plot(loss_train,label='train loss')
    plt.plot(loss_valid,label='valid loss')

    plt.legend(loc='upper right')
    plt.title('epoch: %d '%(epoch))
    plt.pause(.0001)

    print('train loss: ',loss_train[-1])
    print('valid loss: ',loss_valid[-1])
```

I made patch_validation_set like this... but I don't know whether it is right or wrong....I want to make model of predicting age with MRA image as inputs...
and I am so curious about why uniformsampler 'randomly' extracts patches from the volume. I think it might not work well....

**Comment by fepegar** [comment]
> I made patch_validation_set like this... but I don't know whether it is right or wrong

If you want to go through all patches in each image in the validation set, I don't think you need a `Queue`. You could just run the inference using a `GridSampler`, as [shown in the docs](https://torchio.readthedocs.io/data/patch_inference.html).

> I am so curious about why uniformsampler 'randomly' extracts patches from the volume. I think it might not work well

That's how it's designed. It works reasonably well for most applications. If your regions of interest are small, you could use, for example, a `WeightedSampler`

**Comment by 9B8DY6** [comment]
then, what about applying gridsampler to training set for extracting patch from whole volume?
If I try to use WeightedSampler, do I have to make sampling_map for all data? I don't know how to make it....

**Comment by fepegar** [comment]
> then, what about applying gridsampler to training set for extracting patch from whole volume?

The uniform sampler extracts patches from the whole volume with equal probability.

> If I try to use WeightedSampler, do I have to make sampling_map for all data? I don't know how to make it....

A sampling map is just a volume with the same spatial shape as your images. For each voxel, you set the (relative) probability that it will be sampled.

For example, if your image is `[1, 2, 3, 4]`, your sampling map is `[0 1 2 1]` and your patch size is `1` and you extract 1000 patches, you will have approximately 250 2s, 500 3s and 250 4s. Using a uniform sampler is the equivalent of having a sampling map of all ones. If you really, really want to make sure that you use all the patches in the image, I suppose you don't need a queue. You can load the volume and then pass it to a grid sampler to extract the patches. But I think that approach will lead to poorer results.

**Comment by 9B8DY6** [comment]
> For example, if your image is [1, 2, 3, 4], your sampling map is [0 1 2 1] and your patch size is 1 and you extract 1000 patches, 

'extracting 1000 patches'. how can I make it? `max_queue_length=1000`is the right answer?

**Comment by fepegar** [comment]
```python
In [1]: import torchio as tio

In [3]: import torch

In [4]: t = torch.tensor((1, 2, 3, 4)).reshape(1, 1, 1, 4)

In [5]: t
Out[5]: tensor([[[[1, 2, 3, 4]]]])

In [6]: patch_size = 1

In [7]: sampling_map = torch.tensor((0, 1, 2, 1)).reshape(1, 1, 1, 4)

In [8]: subject = tio.Subject(image=tio.ScalarImage(tensor=t), sampling=tio.LabelMap(tensor=sampling_map))

In [9]: sampler = tio.data.WeightedSampler(patch_size, 'sampling')

In [10]: sampler(subject)
Out[10]: <generator object WeightedSampler.__call__ at 0x7f623aca75f0>
```

The variable returned by `sampler(subject)` is a generator, as you can see. You can add an argument to specify the number of patches you want:

```python
In [13]: for patch in sampler(subject, 1000):
    ...:     patches.append(patch)
    ...: 

In [14]: values = [patch.image.data.item() for patch in patches]

In [15]: values.count(1)
Out[15]: 0

In [16]: values.count(2)
Out[16]: 251

In [17]: values.count(3)
Out[17]: 495

In [18]: values.count(4)
Out[18]: 254
```

Have you read all the documentation in the [Patch-based pipelines](https://torchio.readthedocs.io/data/patch_based.html) section?

**Comment by 9B8DY6** [comment]
Of course, I have read it many times...but I am new to both python and deep learning. so I have never thought the solution like you made... I think I need a lot of learning. Thank you

**Comment by fepegar** [comment]
No worries!
